### PR TITLE
Finetune GPT-2

### PR DESCRIPTION
**Pull request by wangkuiyi**
1. Add `*~` to `.gitignore`. Emacs generates such cache files.
2. Add Python function `loss` to `model.py`.
3. Polish `example_gen_test.py` into `example_generate.py` as a better example. It also exports function `generate_text` for `example_finetune.py`.
4. Add `example finetune.py` to show that fine-tuning a GPT-2 model can encode personalized "knowledge." The following example run shows that GPT-2 did not know the names of Yi's dogs at first. Fine-tuning with the training sentence `"Yi Wang has two dogs. One is Joy, the other is Relaxie."` engraves the knowledge into GPT-2 in two iterations.

```
# After install the Python bind of IREE and add iree-jax to PYTHONPATH
08:46 $ python example_finetune.py
Prompt = Yi Wang has two dogs.
Before finetuning, the completion is
 He's a good dog, but he's not a good dog.
After finetuning iteration 0, the completion is
 One is the other is the other is the other is the other is the other is
After finetuning iteration 1, the completion is
 One is Joy, the other is Relaxie.
After finetuning iteration 2, the completion is
 One is Joy, the other is Relaxie.
After finetuning iteration 3, the completion is
 One is Joy, the other is Relaxie.
After finetuning iteration 4, the completion is
 One is Joy, the other is Relaxie.
```

### CODE REVIEW

File: models/gpt2/example_generate.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    kv, x = model.encode(params, kv, prompt, 0, t)
+
+    ret = []
+    ret.append(x.item())
+
+    for _ in range(min(n_new_words, S - nids)):
+        kv, x = model.decode(params, kv, x, t)
+        t = t + 1
+        if x.item() == 50256:
+            break
```
Comment (jpienaar) [inline models/gpt2/example_generate.py:38]: Don't quite know where this constant is coming from. Could you add comment?

**Review by jpienaar — APPROVED** [review]
Nice, thanks! (Not frequent python user, but looks good/can be refined if needed as that's not blocker).

File: models/gpt2/example_generate.py (hunk 1 above)
Comment (wangkuiyi) [reply models/gpt2/example_generate.py:38]: Thaks for the reminder! Done!

**Review by jpienaar — APPROVED** [review]
